include/nil/crypto3/zk/commitments/polynomial/lpc.hpp

```diff
@@ -128,78 +128,50 @@ namespace nil {
-                                    auto& points = this->_points[b_ind][poly_ind];
-                                    BOOST_ASSERT(points.size() == this->_z.get_poly_points_number(b_ind, poly_ind));
-
-                                    std::vector<math::polynomial<value_type>> V_multipliers = this->get_V_multipliers(points);
-
-                                    math::polynomial<value_type> U = this->get_U(b_ind, poly_ind);
-
-                                    math::polynomial<value_type> g_normal(this->_polys[b_ind][poly_ind].coefficients());
-                                    math::polynomial<value_type> Q = g_normal - U;
-
-                                    for (const auto& V_mult: V_multipliers) {
-                                        Q /= V_mult;
-                                    }
-
-                                    math::polynomial_dfs<value_type> Q_dfs(0, _fri_params.D[0]->size());
-                                    Q_dfs.from_coefficients(Q);
-
-                                    if (first) {
-                                        first = false;
-                                        combined_Q = Q_dfs;
-                                    } else {
-                                        combined_Q *= theta;
-                                        combined_Q += Q_dfs;
+                        math::polynomial<value_type> V;
+
+                        auto points = this->get_unique_points();
+                        if constexpr (std::is_same<math::polynomial_dfs<value_type>, PolynomialType>::value ) {
+                            math::polynomial<value_type> combined_Q_normal;
+                            for (auto const &point: points){
+                                bool first = true;
+                                V = {-point, 1};
+                                math::polynomial<value_type> Q_normal;
+                                for(std::size_t i: this->_z.get_batches()){
+                                    for(std::size_t j = 0; j < this->_z.get_batch_size(i); j++){
+                                        auto it = std::find(this->_points[i][j].begin(), this->_points[i][j].end(), point);
+                                        if( it == this->_points[i][j].end()) continue;
+                                        math::polynomial<value_type> g_normal(this->_polys[i][j].coefficients());
+                                        g_normal *= theta_acc;
+                                        Q_normal += g_normal;
+                                        Q_normal -= this->_z.get(i, j, it - this->_points[i][j].begin()) * theta_acc;
+                                        theta_acc *= theta;
                                     }
                                 }
+                                Q_normal = Q_normal / V;
+                                combined_Q_normal += Q_normal;
                             }
+                            combined_Q.from_coefficients(combined_Q_normal);
```

Comment:
Can you remove the code duplication? We are computing this in coefficients form in both places anyway. Only 2 lines are different:

1. math::polynomial<value_type> g_normal(this->_polys[i][j].coefficients()); - You can create a separate function for this, and call from both places. Somethling like 

template<class PolynomialType>
math::polynomial<value_type> get_poly(int i, int j) {...}

2. combined_Q.from_coefficients(combined_Q_normal); - You can only leave this line in "if constexpr (std::is_same<math::polynomial_dfs<value_type>, PolynomialType>::value ) {".